lib/analyzers/two_fer/analyze.rb

```diff
@@ -1,8 +1,275 @@
+      end
+
+      # In the case of:
+      # format("One for %s, one for me.", name)
+      if loc.method_name == :format &&
+         loc.receiver == nil &&
+         loc.arguments[0].type == :str &&
+         loc.arguments[1].type == :lvar
+        approve!(:kernel_format)
+      end
+
+      # In the case of:
+      # "One for %s, one for me." % name
+      if loc.method_name == :% &&
+         loc.receiver.type == :str &&
+         loc.arguments[0].type == :lvar
+        approve!(:string_format)
+      end
+
+      # If we have a one-line method that passes the tests, then it's not
+      # soemthing we've planned for, so let's refer it to a mentor
+      return refer_to_mentor!
+    end
+
+    def check_for_conditional_on_default_argument!
+      loc = first_line_in_method(main_method)
+
+      # If we don't have a conditional, then let's get out of here.
+      #
+      # TODO: We might wnt to refactor this to extract a conditional from the
+      # method rather than insist on it being the first line.
+      return unless loc.type == :if
+
+      # Get the clause of the conditional (ie the bit after the "if" keyword)
+      conditional = extract_conditional_clause(loc)
+
+      # Let's warn about using a better default if they `if name == nil`
+      if is_lvar?(conditional.receiver, :name) &&
+         conditional.first_argument == default_argument
+        disapprove!(:incorrect_default_param)
+      end
+
+      # Same thing but if they do it the other way round, ie `if nil == name`
+      if conditional.receiver == default_argument &&
+         is_lvar?(conditional.first_argument, :name)
+        disapprove!(:incorrect_default_param)
+      end
+    end
+
+    # ###
+    # Analysis helpers
+    # ###
+    def default_argument_is_optimal?
+      default_argument_value == "you"
+    end
+
+    def one_line_solution?
+      main_method.body.line_count == 1
+    end
+
+    def using_string_interpolation?
+      main_method.body.dstr_type?
+    end
+
+    def string_interpolation_has_three_components?
+      #main_method.body.pry
+      main_method.body.children.size == 3
+    end
+
+    # ###
+    # Static analysis helpers
+    # ###
+    def num_lines_in_method(method)
+      method.body.child_nodes.size
+    end
+
+    def first_line_in_method(method)
+      # A begin block signifies multiple lines
+      # so we return the first line.
+      method.body.children.first if main_method.body.type == :begin
+
+      # Without a begin block we just have one line,
+      # so we return the method body, which *is* the first line
+      method.body
+    end
+
+    # Is this an lvar (local variable) with a given name?
+    def is_lvar?(node, name)
+      node.lvar_type? && node.children[0] == name
+    end
+
+    def extract_conditional_clause(loc)
+      loc.children[0]
+    end
+
+    memoize
+    def two_fer_module
+      ExtractModuleOrClass.(root_node, "TwoFer")
+    end
+
+    memoize
+    def main_method
```

Comment:
Total nitpick, but can we call this `target_method` instead of `main_method`? My stupid brain keeps thinking `main` as in the starting point of a program, not the `two_fer` method.

Reply:
I spent ages trying (and failing) to name this. `target_method` is exactly the thing. Thank you!

Reply:
Done in next PR :+1: